libraries/package-deps-hash/src/getRepoState.ts

```diff
@@ -246,6 +246,131 @@ export function applyWorkingTreeState(
+ * @param rootDirectory - The root directory of the Git repository
+ * @param gitPath - The path to the Git executable
+ * @internal
+ */
+function getSubmodulePaths(rootDirectory: string, gitPath?: string): string[] {
+  const submoduleStatusResult: child_process.SpawnSyncReturns<string> = Executable.spawnSync(
+    gitPath || 'git',
+    ['submodule', 'status'],
+    {
+      currentWorkingDirectory: rootDirectory
+    }
+  );
+
+  if (submoduleStatusResult.status !== 0) {
+    ensureGitMinimumVersion(gitPath);
+
+    throw new Error(
+      `git submodule status exited with status ${submoduleStatusResult.status}: ${submoduleStatusResult.stderr}`
+    );
+  }
+
+  // <hash> <path> (heads/main)\n
+  return _parseGitSubmoduleStatus(submoduleStatusResult.stdout);
+}
+
+/**
+ * Parse the state of all git submodule from the output of "git submodule foreach git --no-optional-locks ls-tree -r --full-name HEAD --"
+ *
+ * @param output - The raw output from Git
+ * @internal
+ */
+export function _parseGitSubmoduleForEachGitLsTree(output: string): Record<string, Record<string, string>> {
+  const stateBySubmodulePath: Record<string, Record<string, string>> = {};
+
+  let currentSubmodulePath: string = '';
+
+  // Note: Entering maybe translated to other languages by Git
+  // Entering '<submoduleFolder>'\n
+  // <mode> <type> <newhash>\t<path>\n
+  output.split('\n').forEach((line) => {
+    if (line) {
+      line = line.trim();
+      const submodulePath: string | undefined = line.match(/\s'(.*)'$/)?.[1];
+      if (submodulePath) {
+        currentSubmodulePath = submodulePath;
+        stateBySubmodulePath[currentSubmodulePath] = {};
+      } else {
+        const tabIndex: number = line.indexOf('\t');
+        const filePath: string = line.slice(tabIndex + 1);
+
+        // The newHash will be all zeros if the file is deleted, or a hash if it exists
+        const hash: string = line.slice(tabIndex - 40, tabIndex);
+        if (stateBySubmodulePath[currentSubmodulePath] === undefined) {
+          stateBySubmodulePath[currentSubmodulePath] = {};
+        }
+        stateBySubmodulePath[currentSubmodulePath][filePath] = hash;
+      }
+    }
+  });
+
+  return stateBySubmodulePath;
+}
+
+/**
+ * Augments the state value with modifications that are in git submodules
+ * @param rootDirectory - The root directory of the Git repository
+ * @param state - The current map of git path -> object hash. Will be mutated.
+ * @param gitPath - The path to the Git executable
+ * @internal
+ */
+function applySubmodulesTreeState(rootDirectory: string, state: Map<string, string>, gitPath?: string): void {
+  const submodulePaths: string[] = getSubmodulePaths(rootDirectory, gitPath);
+
+  if (submodulePaths.length === 0) {
+    return;
+  }
+
+  const submoduleLsTreeResult: child_process.SpawnSyncReturns<string> = Executable.spawnSync(
+    gitPath || 'git',
+    ['submodule', 'foreach', 'git', '--no-optional-locks', 'ls-tree', '-r', '--full-name', 'HEAD', '--'],
```

Comment:
You will want to pass `-z` to the `ls-tree` command to handle encoding issues with the file paths